### DISCUSSION TITLE
Getting the Donor ID for use on a page

### DISCUSSION THREAD

**Discussion by Rik44444** [impress-org/givewp#7116] [Developer Q&A]
We are trying to build an extra tab for the Donor Dashboard. For this, we need to use GiveWP's Donor ID as a parameter in on a page. In what way can we get hold of this Donor ID on a Wordpress page? We've noted that there is a unique key in the URL for GiveWP users in the URL, could we get this key somehow? And then via the key get the Donor ID?

**Comment by JoshuaHungDinh** [comment]
Hey @rikk4444, to ensure I understand correctly, could you please share additional details about the intended use case for the donor id, or any specific outcomes that you are looking for? I'd like to get a better understanding what you're looking to achieve with it. Thanks!

**Comment by Rik44444** [comment]
hi @JoshuaHungDinh , yes sure! Thanks for replying. 
We are building an extra tab for the Donor Dashboard. In order to be able to do this, we need some way to get a 'Donor ID' for this page when a user is logged in, in order to be able to present information here from our DB. In our DB, we have used the Donor ID that is exportable via the GiveWP API as the identifying variable for a donor. We are now wondering if there is a snippet or standard bit of code to get that Donor ID when someone is logged in. 
We noticed there is a unique key in the URL for GiveWP users in the URL, which is the same key that is in the Login email that users get from our GiveWP installation. It looks like this: https://giveforgood.world/en/donor-dashboard/?give_nl=AYboojOCn8Vcpl6KrNfe. We wondered if this key might be the best way to get to the Donor ID (eg by extracting if from the URL and then linking the key to the Donor ID in the GiveWP DB), or if there is an easier/more direct way. 
Does that provide the extra info you needed to answer the question? Feel free to ask more if you need to know more!

**Reply by JoshuaHungDinh** [reply]
Hi @Rik44444,

Thank you for the additional information — I believe I have a better understanding. I also checked out the page you sent and it looks fantastic!

I might suggest considering the use of our Donor model, if possible for your implementation. This model provides access to various properties, including the donor ID. A potential approach involves utilizing the get_user_id WordPress function to retrieve the user ID. With the user ID, you can then search for the corresponding donor and obtain their ID.

A short example could be
```
$userId = get_current_user_id();
$donor = \Give\Donors\Models\Donor::whereUserId($userId);
$donorId = $donor->id;
```

**Reply by Rik44444** [reply]
Thanks! And does this also work given the fact that most (all) of our users do not have Wordpress IDs? They login using the 'email access', so as a result they are not shown in our list of Wordpress users.

**Reply by kjohnson** [reply]
I would suggest using `\Give\Donors\Models\Donor::whereEmail()`, if you have the email address, which will support the "email access" donors.

**Comment by kjohnson** [comment]
@Rik44444 we don't have any official resources on extending the Donor Dashboard (that I know of), but I was able to reconstruct a starting point.

You'll register a new "tab" on the donor dashboard during the `init` hook.

The tab is represented as a class, which should extend `\Give\DonorDashboards\Tabs\Contracts\Tab`.

With this tab class you can register routes (api endpoints) and enqueue assets, such as javascript that will render in that tab.

I'll have to do some more digging from there, since this isn't something that is documented, yet.

That said, this should be a good place to start.

```php
<?php

add_action('init', function() {
  (new \App\CustomTab)->registerTab();
});
```

```php
<?php

namespace App;

class CustomTab extends \Give\DonorDashboards\Tabs\Contracts\Tab
{
	public static function id() {
		return 'custom-tab';
	}

	public function routes() {
		return [
			MyCustomTabRoute::class, // Must extend \Give\DonorDashboards\Tabs\Contracts\Route
		];
	}

	public function enqueueAssets()
	{
		// Enqueue assets here
	}
}
```
